src/Tools/IdeCoreBenchmarks/SyntaxListBenchmarks.cs

```diff
@@ -0,0 +1,105 @@
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
+
+using System.IO;
+using BenchmarkDotNet.Attributes;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+
+namespace IdeCoreBenchmarks
+{
+    [MemoryDiagnoser]
+    public class SyntaxListBenchmarks
+    {
+        private SyntaxNode _smallRoot;
+        private SyntaxNode _mediumRoot;
+        private SyntaxNode _largeRoot;
+        private SourceText _smallText;
+        private SourceText _mediumText;
+        private SourceText _largeText;
+
+        private const int IterationCountWalkTree = 1000;
+        private const int IterationCountWalkLines = 10;
+
+        [GlobalSetup]
+        public void GlobalSetup()
+        {
+            (_smallRoot, _smallText) = GetRootAndTextFrom(@"d:\src\Roslyn2\src\Compilers\CSharp\Portable\Parser\AbstractLexer.cs");
+            (_mediumRoot, _mediumText) = GetRootAndTextFrom(@"d:\src\Roslyn2\src\Compilers\CSharp\Portable\Parser\SyntaxParser.cs");
+            (_largeRoot, _largeText) = GetRootAndTextFrom(@"d:\src\Roslyn2\src\Compilers\CSharp\Portable\Parser\LanguageParser.cs");
+        }
+
+        private (SyntaxNode, SourceText) GetRootAndTextFrom(string path)
+        {
+            var textContents = File.ReadAllText(path);
+
+            var text = SourceText.From(textContents);
+            var tree = SyntaxFactory.ParseSyntaxTree(text);
+
+            return (tree.GetCompilationUnitRoot(), text);
+        }
+
+        [Benchmark]
+        public void WalkLines_Small()
+        {
+            WalkLines(_smallText, _smallRoot);
+        }
+
+        [Benchmark]
+        public void WalkTree_Small()
+        {
+            WalkTree(_smallRoot);
+        }
+
+        [Benchmark]
+        public void WalkLines_Medium()
+        {
+            WalkLines(_mediumText, _mediumRoot);
+        }
+
+        [Benchmark]
+        public void WalkTree_Medium()
+        {
+            WalkTree(_mediumRoot);
+        }
+
+        [Benchmark]
+        public void WalkLines_Large()
+        {
+            WalkLines(_largeText, _largeRoot);
+        }
+
+        [Benchmark]
+        public void WalkTree_Large()
+        {
+            WalkTree(_largeRoot);
+        }
+
+        private static void WalkLines(SourceText sourceText, SyntaxNode root)
+        {
+            for (var i = 0; i < IterationCountWalkLines; i++)
+            {
+                foreach (var line in sourceText.Lines)
+                {
+                    foreach (var token in root.DescendantTokens(line.Span))
+                    {
+                    }
+                }
+            }
+        }
+
+        private static void WalkTree(SyntaxNode root)
+        {
+            for (var i = 0; i < IterationCountWalkTree; i++)
+            {
+                foreach (var token in root.DescendantTokens())
+                {
+                }
+            }
+        }
+    }
+}
```

Comment:
could you add:

```c#
 private static void WalkClassification(SourceText sourceText, SyntaxNode root)
        {
            for (var i = 0; i < IterationCountWalkLines; i++)
            {
                foreach (var line in sourceText.Lines)
                {
                    ClassificationTokens(line.Span, root, default);
                }
            }
        }

        private static void ClassificationTokens(TextSpan span, SyntaxNode root, CancellationToken cancellationToken)
        {
            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var stack);
            stack.Push(root);

            var textSpanStart = span.Start;
            var textSpanEnd = span.End;

            while (stack.TryPop(out var current))
            {
                cancellationToken.ThrowIfCancellationRequested();

                // It's ok that we're not pushing in reverse.  The caller (TotalClassificationTaggerProvider) will be
                // sorting the results before doing anything with them.
                foreach (var child in current.ChildNodesAndTokens())
                {
                    if (child.AsNode(out var childNode))
                    {
                        var childSpan = childNode.FullSpan;

                        // If we haven't reached the start of the span we care about, then we can skip this node, going to
                        // the next.  Once we go past that span, we can stop immediately.  Otherwise, we must be
                        // intersecting the span, and we should recurse into this child.
                        if (childSpan.End < textSpanStart)
                            continue;
                        else if (childSpan.Start > textSpanEnd)
                            break;
                        else
                            stack.Push(childNode);
                    }
                    else
                    {
                        ClassifyToken(child);
                    }
                }
            }
        }

        private static void ClassifyToken(SyntaxNodeOrToken child)
        {
        }
```

Reply:
Done! Added data to the PR description with the results for this both with and without the changes in this PR.  (It also had about a 50% improvement, similar to the WalkLines methods)